### PR TITLE
MGMT-6545: fix operator resources

### PR DESCRIPTION
**Pull request by pkliczewski**
This PR introduces new fixture to update cluster and terraform configuration according to requirements needed per specific operator. It is very misleading that the number of nodes ae in both configs and it is not enough to update `workers_count` but we need to do it for duplicated value of `nodes_count`. 

Signed-off-by: Piotr Kliczewski <piotr.kliczewski@gmail.com>

### CODE REVIEW

**Comment by pkliczewski** [comment]
This PR uses changes from https://github.com/openshift/assisted-test-infra/pull/793

**Comment by pkliczewski** [comment]
@YuviGold @tsorya the job failed due to: `Host cannot be installed due to following failing validation(s): All host roles must be assigned to enable OCS in Standard Mode.`. Any clues why not all the hosts get the role assigned?

**Comment by pkliczewski** [comment]
@YuviGold  @tsorya This PR is ready for review

File: discovery-infra/tests/base_test.py [hunk 1]
```diff
@@ -341,7 +343,41 @@ def kube_api_client(self):
...
+            if operators is None:
+                operators = operators_utils.parse_olm_operators_from_env()
+            
+            config.workers_count = operators_utils.resource_param(config.workers_count, consts.OperatorResource.WORKER_COUNT_KEY, operators)
+            config.nodes_count = config.masters_count + config.workers_count
+            config.olm_operators = [operators]
+            
+            return config
+        
+        yield update_config
```
Comment (YuviGold) [inline discovery-infra/tests/base_test.py:383]: Merge those two to get 2 configs and update them both instead

File: discovery-infra/tests/base_test.py [hunk 2]
```diff
@@ -341,7 +343,41 @@ def kube_api_client(self):
 
     @pytest.fixture()
     def kube_api_context(self, kube_api_client):
-        kube_api_context = KubeAPIContext(kube_api_client, clean_on_exit=global_variables.test_teardown)
+        kube_api_context = KubeAPIContext(kube_api_client)
```
Comment (YuviGold) [inline discovery-infra/tests/base_test.py:346]: Seems like a bad rebase. please look into this module for changes that were made and  aren't realted to the OLM changes

File: discovery-infra/tests/base_test.py (hunk 1 above)
Comment (pkliczewski) [reply discovery-infra/tests/base_test.py:383]: My thinking was that we both config may require to be updated in different places but if you think we should be good with one I am happy to update.

File: discovery-infra/tests/base_test.py (hunk 2 above)
Comment (pkliczewski) [reply discovery-infra/tests/base_test.py:346]: yeah, will make sure to keep only changes we need.

File: discovery-infra/tests/base_test.py [hunk 3]
```diff
@@ -40,39 +43,38 @@ def get_nodes(self) -> Callable:
...
+        _net_asset: LibvirtNetworkAssets = nodes_data["net_asset"]
 
         try:
-            if _nodes and global_variables.test_teardown:
+            if global_variables.test_teardown:
                 logging.info('--- TEARDOWN --- node controller\n')
-                _nodes.destroy_all_nodes()
-
+                if _nodes:
+                    _nodes.destroy_all_nodes()
```
Comment (YuviGold) [inline discovery-infra/tests/base_test.py:77]: please take a look at this. Are those changes required or a bad rebase?

Comment (pkliczewski) [reply discovery-infra/tests/base_test.py:77]: I think you made the [changes](https://github.com/openshift/assisted-test-infra/pull/793/files#diff-6162d2a9234339ec7bb76b9037da27056716d9d64db599e223c8d132bc05787d) in this part. Should I restore?

Comment (YuviGold) [reply discovery-infra/tests/base_test.py:77]: I'm pretty sure @eliorerz covered it on https://github.com/openshift/assisted-test-infra/pull/905 . I don't think the change I made is required anymore.

Comment (pkliczewski) [reply discovery-infra/tests/base_test.py]: I restored upstream version, Thanks!

**Comment by YuviGold** [comment]
/approve
/lgtm

Can unhold once system-test-olm is passing (it's a non blocking job)
/hold

File: discovery-infra/tests/test_e2e_install.py [hunk 4]
```diff
@@ -4,25 +4,23 @@
...
 from tests.config import ClusterConfig, TerraformConfig
-from tests.conftest import get_available_openshift_versions
+from tests.conftest import get_available_openshift_versions, get_api_client
 
 
 class TestInstall(BaseTest):
 
     @JunitTestSuite()
-    @pytest.mark.parametrize("openshift_version", get_available_openshift_versions())
+    @pytest.mark.parametrize("openshift_version", sorted(get_available_openshift_versions()))
```
Comment (eliorerz) [inline discovery-infra/tests/test_e2e_install.py:13]: I think it's better for adding the `sorted` on the return value of `get_available_openshift_versions`
```suggestion
    def get_available_openshift_versions() -> List[str]:
        ...
        return sorted(available_versions)
```

File: discovery-infra/tests/test_e2e_install.py [hunk 5]
```diff
@@ -4,25 +4,23 @@
...
     @JunitTestSuite()
-    # TODO: Fix OCS
-    # @pytest.mark.parametrize("olm_operator", get_api_client().get_supported_operators())
-    @pytest.mark.parametrize("olm_operator", ["lso", "cnv"])
-    def test_olm_operator(self, get_nodes, get_cluster, olm_operator):
-        new_cluster = get_cluster(cluster_config=ClusterConfig(olm_operators=[olm_operator]),
-                                  nodes=get_nodes(TerraformConfig(olm_operators=[olm_operator])))
+    @pytest.mark.parametrize("operators", sorted(get_api_client().get_supported_operators()))
+    def test_olm_operator(self, get_nodes, get_cluster, operators, update_olm_config):
+        tf_config, cluster_config = update_olm_config(tf_config=TerraformConfig(), cluster_config=ClusterConfig(), operators=operators)
```
Comment (eliorerz) [inline discovery-infra/tests/test_e2e_install.py:22]: Line is longer than 120 chars

File: discovery-infra/test_infra/controllers/node_controllers/terraform_controller.py [hunk 6]
```diff
@@ -43,18 +43,12 @@ def _get_random_name(cls):
     # TODO move all those to conftest and pass it as kwargs
     # TODO-2 Remove all parameters defaults after moving to new workflow and use config object instead
     def _terraform_params(self, **kwargs):
-        operators = kwargs.get("olm_operators", [])
         params = {
-            "libvirt_worker_memory": operators_utils.resource_param(
-                kwargs.get("worker_memory"), consts.OperatorResource.WORKER_MEMORY_KEY, operators),
-            "libvirt_master_memory": operators_utils.resource_param(
-                kwargs.get("master_memory", 16984), consts.OperatorResource.MASTER_MEMORY_KEY, operators),
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/terraform_controller.py:51]: You have that value on `consts.resources`  
https://github.com/openshift/assisted-test-infra/blob/master/discovery-infra/test_infra/consts/resources.py

File: discovery-infra/test_infra/controllers/node_controllers/terraform_controller.py [hunk 7]
```diff
@@ -64,21 +58,17 @@ def _terraform_params(self, **kwargs):
             "libvirt_dns_records": kwargs.get("dns_records", {}),
             # TODO change to namespace index
             "libvirt_network_if": self.config.net_asset.libvirt_network_if,
-            "libvirt_worker_disk": operators_utils.resource_param(
-                kwargs.get("worker_disk", 21474836480), consts.OperatorResource.WORKER_DISK_KEY, operators),
-            "libvirt_master_disk": operators_utils.resource_param(
-                kwargs.get("master_disk", 128849018880), consts.OperatorResource.MASTER_DISK_KEY, operators),
+            "libvirt_worker_disk": kwargs.get("worker_disk", 26843545601),
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/terraform_controller.py:61]: Why is `worker_disk`, and `master_disk` has different value than the default value we have on `consts.resources`?

File: discovery-infra/test_infra/controllers/node_controllers/terraform_controller.py [hunk 8]
```diff
@@ -64,21 +58,17 @@ def _terraform_params(self, **kwargs):
...
             # TODO change to namespace index
             "libvirt_secondary_network_if": self.config.net_asset.libvirt_secondary_network_if,
             "provisioning_cidr": self.config.net_asset.provisioning_cidr,
             "running": True,
             "single_node_ip": kwargs.get("single_node_ip", ''),
-            "master_disk_count": operators_utils.resource_param(
-                kwargs.get("master_disk_count", 1), consts.OperatorResource.MASTER_DISK_COUNT_KEY, operators),
-            "worker_disk_count": operators_utils.resource_param(
-                kwargs.get("worker_disk_count", 1), consts.OperatorResource.WORKER_DISK_COUNT_KEY, operators),
+            "master_disk_count": kwargs.get("master_disk_count", 1),
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/terraform_controller.py:70]: Here also, values are in `consts.resources`

File: discovery-infra/test_infra/helper_classes/config/nodes_config.py [hunk 9]
```diff
@@ -32,7 +32,6 @@ class BaseTerraformConfig(_BaseConfig, ABC):
     storage_pool_path: str = None
     # running: bool = True
     single_node_ip: str = None
-    olm_operators: List[str] = None
```
Comment (eliorerz) [inline discovery-infra/test_infra/helper_classes/config/nodes_config.py:35]: Have you checked to make sure it's not being used elsewhere (test_infra, QE)?

File: discovery-infra/tests/base_test.py [hunk 10]
```diff
@@ -345,3 +348,27 @@ def kube_api_context(self, kube_api_client):
...
+            tf_config.worker_disk = operators_utils.resource_param(tf_config.worker_disk, consts.OperatorResource.WORKER_DISK_KEY, operators)
+            tf_config.master_disk = operators_utils.resource_param(tf_config.master_disk, consts.OperatorResource.MASTER_DISK_KEY, operators)
+            tf_config.master_disk_count = operators_utils.resource_param(tf_config.master_disk_count, consts.OperatorResource.MASTER_DISK_COUNT_KEY, operators)
+            tf_config.worker_disk_count = operators_utils.resource_param(tf_config.worker_disk_count, consts.OperatorResource.WORKER_DISK_COUNT_KEY, operators)
+
+            cluster_config.workers_count = operators_utils.resource_param(cluster_config.workers_count, consts.OperatorResource.WORKER_COUNT_KEY, operators)
+            cluster_config.nodes_count = cluster_config.masters_count + cluster_config.workers_count
+            cluster_config.olm_operators = [operators]
+
+            return tf_config, cluster_config
```
Comment (eliorerz) [inline discovery-infra/tests/base_test.py:372]: You don't have to return the configurations. It's update function.

File: discovery-infra/tests/base_test.py [hunk 11]
```diff
@@ -345,3 +348,27 @@ def kube_api_context(self, kube_api_client):
 
         with kube_api_context:
             yield kube_api_context
+
+    @pytest.fixture(scope="function")
+    def update_olm_config(self) -> Callable:
```
Comment (eliorerz) [inline discovery-infra/tests/base_test.py:353]: I you remove the return value (few comments next) change the return type annotation to 
```suggestion
    def update_olm_config(self) -> Callable[[TerraformConfig, ClusterConfig, <TypeOfOperators>], None]:
```
The `None` part is the return value type

File: discovery-infra/tests/test_e2e_install.py [hunk 12]
```diff
@@ -4,25 +4,23 @@
...
         new_cluster.start_install_and_wait_for_installed()
 
     @JunitTestSuite()
-    # TODO: Fix OCS
-    # @pytest.mark.parametrize("olm_operator", get_api_client().get_supported_operators())
-    @pytest.mark.parametrize("olm_operator", ["lso", "cnv"])
-    def test_olm_operator(self, get_nodes, get_cluster, olm_operator):
-        new_cluster = get_cluster(cluster_config=ClusterConfig(olm_operators=[olm_operator]),
-                                  nodes=get_nodes(TerraformConfig(olm_operators=[olm_operator])))
+    @pytest.mark.parametrize("operators", sorted(get_api_client().get_supported_operators()))
```
Comment (eliorerz) [inline discovery-infra/tests/test_e2e_install.py:20]: Why not using `api_client` fixture?

File: discovery-infra/tests/test_e2e_install.py (hunk 4 above)
Comment (pkliczewski) [reply discovery-infra/tests/test_e2e_install.py:13]: Ack

File: discovery-infra/tests/test_e2e_install.py (hunk 5 above)
Comment (pkliczewski) [reply discovery-infra/tests/test_e2e_install.py:22]: Will change

File: discovery-infra/test_infra/controllers/node_controllers/terraform_controller.py (hunk 6 above)
Comment (pkliczewski) [reply discovery-infra/test_infra/controllers/node_controllers/terraform_controller.py:51]: OK

File: discovery-infra/test_infra/controllers/node_controllers/terraform_controller.py (hunk 7 above)
Comment (pkliczewski) [reply discovery-infra/test_infra/controllers/node_controllers/terraform_controller.py:61]: I restored old code, haven't checked what the default value is.

File: discovery-infra/test_infra/controllers/node_controllers/terraform_controller.py (hunk 8 above)
Comment (pkliczewski) [reply discovery-infra/test_infra/controllers/node_controllers/terraform_controller.py:70]: I restored old code, haven't checked what the default value is.

File: discovery-infra/test_infra/helper_classes/config/nodes_config.py (hunk 9 above)
Comment (pkliczewski) [reply discovery-infra/test_infra/helper_classes/config/nodes_config.py:35]: I will double check

File: discovery-infra/tests/base_test.py (hunk 10 above)
Comment (pkliczewski) [reply discovery-infra/tests/base_test.py:372]: It makes test code easy if I return them in this way. It makes sense to make to merge both.

File: discovery-infra/tests/base_test.py (hunk 11 above)
Comment (pkliczewski) [reply discovery-infra/tests/base_test.py:353]: OK

File: discovery-infra/tests/test_e2e_install.py (hunk 12 above)
Comment (pkliczewski) [reply discovery-infra/tests/test_e2e_install.py:20]: No good reason :)

File: discovery-infra/test_infra/helper_classes/config/nodes_config.py (hunk 9 above)
Comment (pkliczewski) [reply discovery-infra/test_infra/helper_classes/config/nodes_config.py:35]: According to my findings this is only used from `ClusterConfig`

File: discovery-infra/tests/base_test.py (hunk 11 above)
Comment (pkliczewski) [reply discovery-infra/tests/base_test.py:353]: It is not working due to `TypeError: 'ABCMeta' object is not subscriptable`

File: discovery-infra/tests/test_e2e_install.py (hunk 12 above)
Comment (pkliczewski) [reply discovery-infra/tests/test_e2e_install.py:20]: I explored and see that using fixture in this place won't work: `Fixtures are meant to be called directly, but are created automatically`. We need to run this test in parallel for each operator.

**Review by eliorerz — APPROVED** [review]
LGTM

**Comment by pkliczewski** [comment]
:tada:

**Pull request merged**